src/main/python/camdkit/framework.py

```diff
@@ -233,30 +312,338 @@ def make_json_schema() -> dict:
+      "minimum": 1,
+      "maximum": INT_MAX
+    }
+  
+class NonNegativeRealParameter(Parameter):
+  
+  @staticmethod
+  def validate(value) -> bool:
+    """The parameter shall be a non-negative real number."""
+
+    return isinstance(value, numbers.Real) and value >= 0.0
+
   @staticmethod
   def to_json(value: typing.Any) -> typing.Any:
     return value
 
   @staticmethod
   def from_json(value: typing.Any) -> typing.Any:
-    return int(value)
+    return float(value)
+  
+  @staticmethod
+  def make_json_schema() -> dict:
+    return {
+      "type": "number",
+      "minimum": 0.0,
+    }
+  
+class EnumParameter(StringParameter):
+  allowedValues = []
+  enum_class = None
+
+  @classmethod
+  def __init_subclass__(cls) -> None:
+    if cls.enum_class == None:
+      # Sub-class must specify the related Enum class
+      raise ValueError
+    cls.allowedValues = [e.value for e in cls.enum_class]
+
+  def validate(self, value) -> bool:
+    """The parameter shall be one of the allowed values."""
+    return str(value) in self.allowedValues
+
+  def make_json_schema(self) -> dict:
+    return {
+      "type": "string",
+      "enum": self.allowedValues
+    }
+  
+class TimingModeParameter(EnumParameter):
+  sampling = Sampling.REGULAR
+  enum_class = TimingMode
+  
+class TimestampParameter(Parameter):
+  sampling = Sampling.REGULAR
+
+  @staticmethod
+  def validate(value) -> bool:
+    """
+    The parameter shall contain valid number of seconds, nanoseconds and optionally
+    attoseconds elapsed since the start of the epoch.
+    """
+    if not isinstance(value, Timestamp):
+      return False
+    if not (isinstance(value.seconds, int) and value.seconds >= 0 and value.seconds <= UINT48_MAX):
+      return False
+    if not (isinstance(value.nanoseconds, int) and value.nanoseconds >= 0 and value.nanoseconds <= UINT_MAX):
+      return False
+    if not (isinstance(value.attoseconds, int) and value.attoseconds >= 0 and value.attoseconds <= UINT_MAX):
+      return False
+    return True
+
+  @staticmethod
+  def to_json(value: typing.Any) -> typing.Any:
+    return dataclasses.asdict(value)
+
+  @staticmethod
+  def from_json(value: typing.Any) -> typing.Any:
+    return Timestamp(**value)
 
   @staticmethod
   def make_json_schema() -> dict:
     return {
-      "type": "integer",
-      "minimum": 1,
-      "maximum": 2147483647
+      "type": "object",
+      "additionalProperties": False,
+      "properties": {
+        "seconds": {
+          "type": "integer",
+          "minimum": 0,
+          "maximum": UINT48_MAX
+        },
+        "nanoseconds": {
+          "type": "integer",
+          "minimum": 0,
+          "maximum": UINT_MAX
+        },
+        "attoseconds": {
+          "type": "integer",
+          "minimum": 0,
+          "maximum": UINT_MAX
+        }
+      },
+      "required": ["seconds", "nanoseconds"]
+    }
+
+
+class TimecodeParameter(Parameter):
+  sampling = Sampling.REGULAR
```

Comment:
`sampling` is set in `model`. The idea is that any parameter can be used either statically or dynamically.